src/types/types.ts

```diff
@@ -169,21 +170,61 @@ class PDAConfigurationMemento implements IConfigurationMemento {
     }
 }
 
-export interface ISimulation {
+export interface ISimulation{
     automaton: IAutomaton;
     configuration: IAutomatonConfiguration;
 
-    commandHistory: RunCommand<unknown>[];
-    executeCommand<T>(command: RunCommand<T>): void; // if (command.execute()) { commandHistory.push(command); }
+    commandHistory: RunCommand[];
+    executeCommand(command: RunCommand): void; // if (command.execute()) { commandHistory.push(command); }
     undo(): void; // command = commandHistory.pop(); command.undo();
 
     run(): void;
 }
 
-export abstract class RunCommand<T = void> {
+export class Simulation implements ISimulation {
+    automaton: IAutomaton;
+    configuration: IAutomatonConfiguration;
+    commandHistory: (RunCommand)[];
+
+    constructor(_automaton: IAutomaton, _configuration: IAutomatonConfiguration){
+            this.automaton = _automaton;
+            this.configuration = _configuration;
+            this.commandHistory = [];
+    }
+
+    executeCommand(command: RunCommand): void {
+        if (command.execute()) {
+             this.commandHistory.push(command); 
+            }
+        // maybe return error if fails to execute
+    }
+    undo(): void {
+        const command = this.commandHistory.pop();
+        if (command === undefined) return; // maybe error message when we try undo on empty history
+            else command.undo();
+    }
+
+    // simulates the entire run on the word in configuration (or the remaining word) and returns true if the
+    // last state is accepting and false if it isn't
+    run(): boolean {
+
+        while(this.configuration.remainingInput.length>0){
+            const nextCommand = new NextStepCommand(this);
+            if (nextCommand.execute()) {
+                this.commandHistory.push(nextCommand); 
+               }
+        }
+        this.automaton.finalStateIds.forEach(element  => {if(element===this.configuration.stateId) return true; });
+        return false;
```

Comment:
This always returns `false`. The `return true` part from your `forEach` only returns `true` from the fuction that you pass as a parameter to `forEach`, not from the outer function.  Something like this could work instead:
```suggestion
        return this.automaton.finalStateIds.some(finalStateId => this.configuration.stateId === finalStateId);
```

Reply:
aha jasné dík